projects/fluid_simulation/sim.cpp

```diff
@@ -0,0 +1,435 @@
+    mDrawColorTexture    = std::make_unique<Texture>(this, "draw color", GetApp()->GetWindowWidth(), GetApp()->GetWindowHeight(), GetApp()->GetSwapchain()->GetColorFormat());
+    mDyeTexture[0]       = std::make_unique<Texture>(this, "dye[0]", dyeRes.x, dyeRes.y, kRGBA);
+    mDyeTexture[1]       = std::make_unique<Texture>(this, "dye[1]", dyeRes.x, dyeRes.y, kRGBA);
+    mVelocityTexture[0]  = std::make_unique<Texture>(this, "velocity[0]", simRes.x, simRes.y, kRG);
+    mVelocityTexture[1]  = std::make_unique<Texture>(this, "velocity[1]", simRes.x, simRes.y, kRG);
+
+    InitBloomTextures();
+    InitSunraysTextures();
+}
+
+void FluidSimulation::InitBloomTextures()
+{
+    ppx::int2 res = GetResolution(mConfig.bloomResolution);
+    mBloomTexture = std::make_unique<Texture>(this, "bloom", res.x, res.y, kRGBA);
+    PPX_ASSERT_MSG(mBloomTextures.empty(), "Bloom FBOs already initialized");
+    for (int i = 0; i < mConfig.bloomIterations; i++) {
+        uint32_t width  = res.x >> (i + 1);
+        uint32_t height = res.y >> (i + 1);
+        if (width < 2 || height < 2)
+            break;
+        mBloomTextures.emplace_back(std::make_unique<Texture>(this, std::string("bloom frame buffer[") + std::to_string(i) + "]", width, height, kRGBA));
+    }
+}
+
+void FluidSimulation::InitSunraysTextures()
+{
+    ppx::int2 res       = GetResolution(mConfig.sunraysResolution);
+    mSunraysTexture     = std::make_unique<Texture>(this, "sunrays", res.x, res.y, kR);
+    mSunraysTempTexture = std::make_unique<Texture>(this, "sunrays temp", res.x, res.y, kR);
+}
+
+void FluidSimulation::AddTextureToInitialize(Texture* texture)
+{
+    mTexturesToInitialize.push_back(texture);
+}
+
+void FluidSimulation::GenerateInitialSplat()
+{
+    for (const auto& texture : mTexturesToInitialize) {
+        ScheduleDR(mColor->GetDR(texture, ppx::float4(0.0, 0.0f, 0.0f, 1.0f)));
+    }
+    mTexturesToInitialize.clear();
+
+    MultipleSplats(Random().UInt32() % 20 + 5);
+    Render();
+}
+
+ppx::float3 FluidSimulation::HSVtoRGB(ppx::float3 hsv)
+{
+    float       h = hsv[0];
+    float       s = hsv[1];
+    float       v = hsv[2];
+    float       i = floorf(h * 6);
+    float       f = h * 6 - i;
+    float       p = v * (1 - s);
+    float       q = v * (1 - f * s);
+    float       t = v * (1 - (1 - f) * s);
+    ppx::float3 rgb(0, 0, 0);
+
+    switch (static_cast<int>(i) % 6) {
+        case 0: rgb = ppx::float3(v, t, p); break;
+        case 1: rgb = ppx::float3(q, v, p); break;
+        case 2: rgb = ppx::float3(p, v, t); break;
+        case 3: rgb = ppx::float3(p, q, v); break;
+        case 4: rgb = ppx::float3(t, p, v); break;
+        case 5: rgb = ppx::float3(v, p, q); break;
+        default: PPX_ASSERT_MSG(false, "Invalid HSV to RGB conversion");
+    }
+
+    return rgb;
+}
+
+ppx::float3 FluidSimulation::GenerateColor()
+{
+    ppx::float3 c = HSVtoRGB(ppx::float3(Random().Float(0, 1), 1, 1));
+    c.r *= 0.15;
+    c.g *= 0.15;
+    c.b *= 0.15;
```

Comment:
Curious why this? Note that this does not "desaturate" RGB as one would think

Reply:
No clue.  Again, taken straight out of the original code.  I'll play with these settings later.